File: 백준/Gold/19760. Операция ＜＜Перестановка＞＞/Операция ＜＜Перестановка＞＞.py
import sys
def print(*args, sep=" ", end="\n"): return sys.stdout.write(sep.join(map(str, args)) + end)
def input(): return sys.stdin.readline().rstrip()

from collections import deque

def main():
    n, m = map(int, input().split())
    deg = [0]*(n+1)
    e = [[] for _ in range(n+1)]
    for i in range(1, m+1):
        x, y = map(int, input().split())
        e[x].append((y, i))
        deg[y] += 1
    dq = deque()
    for i in range(1, n+1):
        if not deg[i]: dq.append(i)
    ans = 0
    while dq:
        if len(dq) > 1: return int(not print(-1))
        cur = dq.popleft()
        for nxt, idx in e[cur]:
            deg[nxt] -= 1
            if not deg[nxt]: 
                dq.append(nxt)
                ans = max(ans, idx)
    print(ans)
    return 0
if __name__ == '__main__':
    sys.exit(main())

File: 백준/Gold/19760. Операция ＜＜Перестановка＞＞/Операция ＜＜Перестановка＞＞.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, m; cin >> n >> m;
    vector<vector<pii>> e(n+1);
    vector<int> deg(n+1);
    for(int i = 1; i <= m; i++){
        int x, y; cin >> x >> y;
        e[x].push_back({y, i});
        deg[y]++;
    }
    queue<int> q;
    for(int i = 1; i <= n; i++) if(!deg[i]) q.push(i);
    int ans = 0;
    while(q.size()){
        if(q.size() > 1) return !(cout << -1);
        int cur = q.front(); q.pop();
        for(auto [nxt, idx] : e[cur]){
            deg[nxt]--;
            if(!deg[nxt]){
                q.push(nxt);
                ans = max(ans, idx);
            }
        }
    }
    cout << ans;
    return 0;
}